File: HW5-advection-transient/Python/transient_compacting_column.py
import sys
sys.path.insert(1, '../../HW1-Numerics/Python/')
sys.path.insert(1, '../../HW2-BC_LBVP/Python/')
sys.path.insert(1, '../../HW3-Hetero-Fluxes-NeuBC/Python/')
sys.path.insert(1, '../../HW4-1DMelt-Migration/Python/')

# import python libraries
import numpy as np
import scipy.sparse as sp
import matplotlib.pyplot as plt

from classfun import *
from build_gridfun2D import build_grid 
from build_opsfun2D import build_ops
from build_bndfun_optimized import build_bnd
from solve_lbvpfun_optimized import solve_lbvp
from comp_fluxfun_gen import comp_flux_gen
from comp_mean_matrix import comp_mean
from solve_Helmholtz import solve_Helmholtz
from solve_Poisson import solve_Poisson
from flux_upwind import flux_upwind


def plotting(phi,hD,uD,qD,vD,pD,zc,zf,i,dt):
    #Plotting the solution
    fig, (ax1, ax2, ax3, ax4) = plt.subplots(1, 4,figsize=(10,7))
    fig.suptitle(f'Modified Helmholtz and Poisson problems at t= {(i+1)*dt}',)
    
    ax1.plot(phi,zc,label='$\phi_D$')
    ax1.set_xlabel(r'$\phi_D$')
    ax1.set_ylabel(r'$z_D$')
    
    ax2.plot(hD,zc,label=r'$h_D$')
    ax2.plot(uD,zc,label=r'$u_D$')
    ax2.set_xlabel(r'$h_D,u_D$')
    ax2.legend(loc='best')
    
    ax3.plot(pD,zc)
    ax3.set_xlabel(r'$p_D$')
    
    ax4.plot(qD,zf,label=r'$q_D$')
    ax4.plot(vD,zf,label=r'$v_D$')
    ax4.set_xlabel(r'$q_D,v_D$')
    ax4.legend(loc='best')
    plt.tight_layout()

# Parameters
n_exp = 3
m_exp = 1
phic  = 1e-3

# Dimensionless parameters
tmax  = 10
HD    = 50
Nt    = tmax*100
theta = 1 #theta: 1 = forward Euler (explicit), 0.5 = Crank Nicolson, 0 = backward Euler

# Build grid and operator
Grid.xmin = 0; Grid.xmax = HD; Grid.Nx = 300
Grid        = build_grid(Grid)
zc          = Grid.xc
zf          = Grid.xf
[D,G,C,I,M] = build_ops(Grid)
Gamma       = np.zeros((Grid.N,1))

############################################################
#1. Modified Helmholtz equation 
############################################################
# Build boundary conditions
BC.h.dof_dir   = np.array([])
BC.h.dof_f_dir = np.array([]) 
BC.h.g         = np.array([]) 
BC.h.dof_neu   = np.array([Grid.dof_xmin[0],Grid.dof_xmax[0]])
BC.h.dof_f_neu = np.array([Grid.dof_f_xmin[0],Grid.dof_f_xmax[0]])
BC.h.qb        = np.array([[0],[0]])
[B_h,N_h,fn_h] = build_bnd(BC.h, Grid, I)

############################################################
#2. Poisson equation 
############################################################
# Build boundary conditions
BC.u.dof_dir   = np.array([Grid.dof_xmin[0]])
BC.u.dof_f_dir = np.array([Grid.dof_f_xmin[0]]) 
BC.u.g         = np.array([[HD]]) 
BC.u.dof_neu   = np.array([Grid.dof_xmax[0]])
BC.u.dof_f_neu = np.array([Grid.dof_f_xmax[0]])
BC.u.qb        = np.array([[0]])
[B_u,N_u,fn_u] = build_bnd(BC.u, Grid, I)

############################################################
#3. Porosity evolution
############################################################
# Build boundary conditions
BC.phi.dof_dir   = np.array([])
BC.phi.dof_f_dir = np.array([]) 
BC.phi.g         = np.array([]) 
BC.phi.dof_neu   = np.array([Grid.dof_xmin[0],Grid.dof_xmax[0]])
BC.phi.dof_f_neu = np.array([Grid.dof_f_xmin[0],Grid.dof_f_xmax[0]])
BC.phi.qb        = np.array([[0],[0]])
[B_phi,N_phi,fn_phi]= build_bnd(BC.phi, Grid, I)

#Initial condition
phi = np.ones(Grid.Nx)
dt  = tmax/Nt

#Steady state
#phi = 1 + 0.1*np.cos(2*np.pi/HD*zc);
#[hD,pD,qD] = solve_Helmholtz(D,G,I,M,phi,n_exp,m_exp,Grid,B_h,N_h,fn_h,BC.h,zc,Gamma)
#[uD,vD]    = solve_Poisson(D,G,I,phi,m_exp,pD,Grid,B_u,N_u,fn_u,BC.u)


for i in range(0,Nt):

    if (i+1)%100 ==0: plotting(phi,hD,uD,qD,vD,pD,zc,zf,i,dt)

    [hD,pD,qD] = solve_Helmholtz(D,G,I,M,phi,n_exp,m_exp,Grid,B_h,N_h,fn_h,BC.h,zc,Gamma)

    [uD,vD]    = solve_Poisson(D,G,I,phi,m_exp,pD,Grid,B_u,N_u,fn_u,BC.u)
    print(f'i = {i},t = {dt*i},Summing the porosity = {np.sum(phi)}')
    
    A  = flux_upwind(vD,Grid)
    P  = sp.spdiags(pD, 0, Grid.Nx, Grid.Nx)
    L  = phic * D @ A - P
    IM = I + dt * (1-theta) * L
    EX = I - dt * theta * L
    phi= solve_lbvp(IM, EX @ phi, B_phi, BC.phi.g, N_phi)
    
    if np.max(phi*phic) > 1 or np.min(phi*phic) < 0 : break

File: HW5-advection-transient/MATLAB/transient_compacting_column.m
clear 
%% Define physical parameters
n = 3;
m = 1;  % implementation only works for m
phic = 1e-3;

%% Dimensionless parameters
tmax = 10;
Z    = 50;

%% Define Grid and build operators
Grid.xmin = 0;
Grid.xmax = Z;
Grid.Nx   = 300;
Nt = tmax*100;
theta = 1; % forward  Euler (explicit)

Grid = build_grid(Grid);
[D,G,C,I,M] = build_ops(Grid);

%% Define boundary conditions
% 1) mod. Helmholtz eqn.
BC.h.dof_dir   = [];
BC.h.dof_f_dir = [];
BC.h.g         = [];
BC.h.dof_neu   = [];
BC.h.dof_f_neu = [];
BC.h.qb        = [];
[B_h,N_h,fn_h] = build_bnd(BC.h,Grid,I);

% 2) Poisson eqn.
BC.u.dof_dir   = [Grid.dof_xmin];
BC.u.dof_f_dir = [Grid.dof_f_xmin];
BC.u.g         = [0];
BC.u.dof_neu   = [];
BC.u.dof_f_neu = [];
BC.u.qb        = [];
[B_u,N_u,fn_u] = build_bnd(BC.u,Grid,I);

% 3) Advection eqn.
BC.phi.dof_dir   = [];
BC.phi.dof_f_dir = [];
BC.phi.g         = [];
BC.phi.dof_neu   = [];
BC.phi.dof_f_neu = [];
BC.phi.qb        = [];
[B_phi,N_phi,fn_phi] = build_bnd(BC.phi,Grid,I);

%% Initial condition
phi = ones(Grid.N,1);
%phiD = phi;
dt = tmax/Nt;

%% Initialize video
myVideo = VideoWriter('myVideoFile'); %open video file
myVideo.FrameRate = 10;  %can adjust this, 5 - 10 works well for me
open(myVideo)

for i = 1:Nt
    % Flow calculations
    [hD,pD,qD] = solve_Helmholtz(D,G,I,M,phi,n,m,Grid,B_h,N_h,fn_h,BC.h,Grid.xc,zeros(Grid.Nx,1));
    [uD,vD] = solve_Poisson(D,G,I,phi,m,pD,Grid,B_u,N_u,fn_u,BC.u);
    if mod(i,10) == 0
        clf()
        plot_solution(Grid,phi,hD,uD,qD,vD,pD)
        string = "t ="+num2str(i*dt);
        title(string);
        fprintf('i = %d: t= %3.2f V = %3.7f;\n',i,dt*i,sum(phi));
        frame = getframe(gcf); %get frame
        writeVideo(myVideo, frame);
    end
    % Transport calculations
    A = flux_upwind(vD,Grid);
    P = spdiags(pD,0,Grid.Nx,Grid.Nx);    
    L = phic * D * A - P;
    IM = I + dt*(1-theta)*L;
    EX = I - dt*theta*L;
    phi = solve_lbvp(IM , EX * phi, B_phi, BC.phi.g, N_phi); 
    %phi  = phiD; 
    if max(phi*phic) > 1 || min(phi*phic) < 0; error('Porosity outside  [0 1].\n'); end
end

%{
function [] = plot_solution(Grid,phiD,hD,uD,qD,vD,pD)
subplot 141
plot(phi,Grid.xc,'linewidth',2), hold on
% xlim([0 2])
xlabel '\phi_D'
ylabel 'z_D'

subplot 142
plot(hD,Grid.xc,'linewidth',2), hold on
plot(uD,Grid.xc,'linewidth',2)
% xlim(Grid.xmax*[-5 1])
xlabel 'h_D and u_D'
ylabel 'z_D'
legend('h_D','u_D')

subplot 143
plot(qD,Grid.xf,'linewidth',2), hold on
plot(vD,Grid.xf,'linewidth',2)
% xlim(20*[-1 1])
xlabel 'q_D and v_D'
ylabel 'z_D'
legend('q_D','v_D')

subplot 144
plot(pD,Grid.xc,'linewidth',2), hold on
xlim(2*[-1 1])
xlabel 'p_D'
ylabel 'z_D'
drawnow
end
%}
function [] = plot_solution(Grid,phiD,hD,uD,qD,vD,pD)
subplot 141
plot(phiD,Grid.xc,'linewidth',2), hold on
% xlim([0 2])
xlabel '\phi_D'
ylabel 'z_D'

subplot 142
plot(hD,Grid.xc,'linewidth',2), hold on
plot(uD,Grid.xc,'linewidth',2)
% xlim(Grid.xmax*[-5 1])
xlabel 'h_D and u_D'
ylabel 'z_D'
legend('h_D','u_D')

subplot 143
plot(qD,Grid.xf,'linewidth',2), hold on
plot(vD,Grid.xf,'linewidth',2)
% xlim(20*[-1 1])
xlabel 'q_D and v_D'
ylabel 'z_D'
legend('q_D','v_D')

subplot 144
plot(pD,Grid.xc,'linewidth',2), hold on
xlim(2*[-1 1])
xlabel 'p_D'
ylabel 'z_D'
drawnow
end